dask/highlevelgraph.py

```diff
@@ -78,6 +81,30 @@ def get_output_keys(self) -> Set:
         """
         return self.keys()
 
+    def expand_annotations(self) -> Mapping[str, Any]:
+        """Expands associated annotations into a :code:`{key: annotation}`, mapping
+
+        Annotations associated with the Layer are expanded over the set
+        of output keys. Value or literal annotations are copied to each key
+        while callable annotations are called with each key.
+
+        Returns
+        -------
+        annotations : dict or None
+            :code:`{key: annotation}` mapping
+        """
+        if not self.annotations:
+            return None
+
+        annotations = self.annotations
+
+        return {
+            stringify(k): {
+                a: v(k) if callable(v) else v for a, v in annotations.items()
+            }
+            for k in self.get_output_keys()
+        }
```

Comment:
Taking a look at a test currently I see that the form of this looks like the following:

```python
(Pdb) pp annotations
{"('ones-dc03ec60d76a92183714569c316632e9', 0)": {'priority': 15,
                                                  'retries': 2,
                                                  'workers': ('tcp://127.0.0.1:46115',)},
 "('ones-dc03ec60d76a92183714569c316632e9', 1)": {'priority': 15,
                                                  'retries': 2,
                                                  'workers': ('tcp://127.0.0.1:46115',)}}
```

In the future I anticipate that iterating through all keys for all types of built-in annotations (priority, workers, ...) will become a bottleneck and that we'll want to pass up scalar annotations (strings, numbers) (like `dask.annotate(foo=1, bar="baz")` and expand them on the scheduler side, probably in `update_graph_hlg`.  We could do this now, or we could wait.  However, if we want to wait then we should anticipate that future change, and maybe consider transposing the key/annotation relationship here and arrange this as follows:

```python
{
    "priority": {"(x, 0)": 1, "(x, 1)": 1},
    "retries": {"(x, 0)": 1, "(x, 1)": 1},
}
```

I think that ideally though, expand annotations looks through each of the values in self.annotations and does the following:

1.  If it's a scalar, do nothing
2.  If it's a dict, stringify keys
3.  If it's a callable, then call it and stringify keys

Thoughts?

Reply:
I was thinking of 

```python
def pack_annotations(self):
    if self.annotations is None:
        return None
    elif any(callable(v) for v in self.annotations.values()):
        return self.expand_annotations()
   else:
       return tuple(self.annotations.items())
```

This way, unpackers can check for tuple or dict and continue with annotation expansion on if it's a tuple.

Reply:
I'm happy to put this in now.

Reply:
What if one of the annotations is a callable and one is an integer?

I'm thinking more like the following:

```python
out = {}

for annotation, value in self.annotations.values():
    if callable(value):
        out[annotation] = {stringify(k): callable(k[1:]) for k in self.keys}
    elif isinstance(value, dict) and not any(k in self for k in value):
        out[annotation] = keymap(stringify, value)
    else:
        out[annotation] = value
```

And then on the scheduler side we would expand the scalar values

Dicts are a bit strange.  I added a key check a bit in order to handle cases like `{"gpu": 1}`

Reply:
It occurred to me that this kind of transposed structure is ideal for the existing taxonomy as instead of looping over each annotation per-task to update priority for e.g., the following is possible

```python
annotations = {
    "priority": {"(x, 0)": 1, "(x, 1)": 1},
    "retries": {"(x, 0)": 1, "(x, 1)": 1},
}

priorities.update(annotations['priority'])
retries.update(annotations["retries"]
```

I'm keen to experiment with the above idea.

> Dicts are a bit strange. I added a key check a bit in order to handle cases like {"gpu": 1}

I would not support the second statement in the if-else ladder (commented out below)

```python
out = {}

for annotation, value in self.annotations.values():
    if callable(value):
        out[annotation] = {stringify(k): callable(k) for k in self.keys}
    # elif isinstance(value, dict) and not any(k in self for k in value):
    #    out[annotation] = keymap(stringify, value)
    else:
        out[annotation] = value
```

because there's a cart before the horse situation here ("a" needs to be A.name) which makes it very tricky to express the second statement:

```python
with dask.annotate(workers={("a", 0): [a.address], ("a", 1): [b.address]}):
   A = da.ones((10, 10), chunks=(2, 2))
```

I've never considered directly specifying per-task annotations in the layer annotation because of the above, so the code currently considers dicts to be single "values".

Reply:
OK, dropping per-key annotations specified by users seems like an easy path forward to me

Reply:
I've implemented the transposed annotation packing and unpacking.